File: apps/backend/schemas/webhook.py
from pydantic import BaseModel
from enum import Enum
from typing import Dict, List, Optional


class WebhookSchema(BaseModel):
    url: str
    headers: Optional[Dict[str, str]] = None


class ProcessingStatus(str, Enum):
    STARTED = "started"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    ERROR = "error"
    BATCH_PROGRESS = "batch_progress"
    BATCH_COMPLETED = "batch_completed"


class ProcessingType(str, Enum):
    INDIVIDUAL = "individual"
    BATCH = "batch"
    TERMINATE = "terminate"


class ProcessingPhase(str, Enum):
    DOWNLOAD = "download"
    PDF_TO_JSON = "pdf_to_json"
    JSON_TO_GRAPH = "json_to_graph"


class EnvelopeStatus(str, Enum):
    PENDING = "pending"
    COMPLETED = "completed"
    FAILED = "failed"


# Individual envelope processing models
class DocumentProgress(BaseModel):
    current_document: str
    completed: int
    total: int
    percentage: float


class IndividualStartedMessage(BaseModel):
    type: ProcessingType = ProcessingType.INDIVIDUAL
    status: ProcessingStatus = ProcessingStatus.STARTED
    envelope_id: str
    total_documents: int
    phase: ProcessingPhase


class IndividualProgressMessage(BaseModel):
    type: ProcessingType = ProcessingType.INDIVIDUAL
    status: ProcessingStatus = ProcessingStatus.IN_PROGRESS
    envelope_id: str
    progress: DocumentProgress
    phase: ProcessingPhase


class IndividualCompletedMessage(BaseModel):
    type: ProcessingType = ProcessingType.INDIVIDUAL
    status: ProcessingStatus = ProcessingStatus.COMPLETED
    envelope_id: str
    files: List[str]
    phase: ProcessingPhase


class IndividualErrorMessage(BaseModel):
    type: ProcessingType = ProcessingType.INDIVIDUAL
    status: ProcessingStatus = ProcessingStatus.ERROR
    envelope_id: str
    error: str
    phase: ProcessingPhase


# Batch processing models
class EnvelopeStatusInfo(BaseModel):
    total_documents: int
    completed_documents: int
    status: EnvelopeStatus


class OverallProgress(BaseModel):
    completed_envelopes: int
    total_envelopes: int
    completed_documents: int
    total_documents: int
    percentage: float


class CurrentEnvelopeProgress(BaseModel):
    id: str
    current_document: str
    completed: int
    total: int


class BatchProgressMessage(BaseModel):
    type: ProcessingType = ProcessingType.BATCH
    status: ProcessingStatus = ProcessingStatus.BATCH_PROGRESS
    overall_progress: OverallProgress
    current_envelope: Optional[CurrentEnvelopeProgress] = None
    envelope_statuses: Dict[str, EnvelopeStatusInfo]
    phase: ProcessingPhase


class BatchCompletedMessage(BaseModel):
    type: ProcessingType = ProcessingType.BATCH
    status: ProcessingStatus = ProcessingStatus.BATCH_COMPLETED
    overall_progress: OverallProgress
    envelope_statuses: Dict[str, EnvelopeStatusInfo]
    phase: ProcessingPhase


class TerminateMessage(BaseModel):
    type: ProcessingType = ProcessingType.TERMINATE
    terminate: bool = False  # True when this is the final message

File: apps/frontend/src/types/webhook.ts
export type ProcessingPhase =
  | 'download'
  | 'pdf_to_json'
  | 'json_to_graph'
  | 'terminate';

// Common types
export type ProcessingStatus =
  | 'started'
  | 'in_progress'
  | 'completed'
  | 'error'
  | 'batch_progress'
  | 'batch_completed'
  | 'terminated';

export type ProcessingType = 'individual' | 'batch';

// Individual envelope progress types
interface IndividualProgress {
  current_document: string;
  completed: number;
  total: number;
  percentage: number;
}

// Base interface for stored messages (with timestamp)
interface StoredMessageBase {
  timestamp: string;
}

// Individual message types
export interface IndividualStartedMessage extends StoredMessageBase {
  type: 'individual';
  status: 'started';
  envelope_id: string;
  total_documents: number;
  phase: ProcessingPhase;
}

export interface IndividualProgressMessage extends StoredMessageBase {
  type: 'individual';
  status: 'in_progress';
  envelope_id: string;
  progress: IndividualProgress;
  phase: ProcessingPhase;
}

export interface IndividualCompletedMessage extends StoredMessageBase {
  type: 'individual';
  status: 'completed';
  envelope_id: string;
  files: string[];
  phase: ProcessingPhase;
}

export interface IndividualErrorMessage extends StoredMessageBase {
  type: 'individual';
  status: 'error';
  envelope_id: string;
  error: string;
  phase: ProcessingPhase;
}

// Batch progress types
interface OverallProgress {
  completed_envelopes: number;
  total_envelopes: number;
  completed_documents: number;
  total_documents: number;
  percentage: number;
}

interface EnvelopeStatus {
  total_documents: number;
  completed_documents: number;
  status: 'pending' | 'completed';
}

interface CurrentEnvelopeInfo {
  id: string;
  current_document: string;
  completed: number;
  total: number;
}

export interface BatchProgressMessage extends StoredMessageBase {
  type: 'batch';
  status: 'batch_progress';
  overall_progress: OverallProgress;
  current_envelope?: CurrentEnvelopeInfo;
  envelope_statuses: Record<string, EnvelopeStatus>;
  phase: ProcessingPhase;
}

export interface BatchCompletedMessage extends StoredMessageBase {
  type: 'batch';
  status: 'batch_completed';
  overall_progress: OverallProgress;
  envelope_statuses: Record<string, EnvelopeStatus>;
  phase: ProcessingPhase;
}

interface TerminateMessage extends StoredMessageBase {
  type: 'terminate';
  status: 'terminated';
  phase: 'terminate';
  terminate: boolean;
}

// Union type for all webhook messages
export type StoredWebhookMessage =
  | IndividualStartedMessage
  | IndividualProgressMessage
  | IndividualCompletedMessage
  | IndividualErrorMessage
  | BatchProgressMessage
  | BatchCompletedMessage
  | TerminateMessage;

// For incoming messages (without timestamp)
export type WebhookMessage = Omit<StoredWebhookMessage, 'timestamp'>;

// Type for tracking overall processing state
export interface ProcessingState {
  isComplete: boolean;
  progress: number;
  messages: StoredWebhookMessage[];
  currentPhase: ProcessingPhase;
}
